### DISCUSSION TITLE
JuiceStreams show incorrect snapping value

### DISCUSSION THREAD

**Discussion by UnluckyCroco** [ppy/osu#29678] [Q&A] [answered]
### Type

Game behaviour

### Bug description

To clarify with snapping value I mean the box on the right which indicates the distance between the current object and the next (at least that's what it did in stable)

It seems JuiceStreams calculate snapping value based on the start of the JuiceStream (instead of the end). 

In the video below the first JuiceStream should have a snapping value of about 0x, while it has a value of 0.69x. The second JuiceStream should have a value of about 3x, but instead has a value of 0.17x.

### Screenshots or videos

https://github.com/user-attachments/assets/4cfb2538-38b4-427d-914a-be17a8e9d7a6

### Version

2024.817.0-lazer

### Logs

[compressed-logs.zip](https://github.com/user-attachments/files/16829758/compressed-logs.zip)

**Comment by peppy** [comment]
I don't quite understand what you expect from the value, and how you would use this. Like you say it should be "0x" but what does this even mean to you?

**Reply by UnluckyCroco** [reply]
0x would mean the next object is directly above the current one, meaning no additional movement is required to hit it. Basically a higher value would mean higher distance between 2 objects. This value is important to us because we often adjust the Slider Velocity Multiplier so that a 1.00x is perfectly walkable, and gauge how far to place objects based on that.

So the reason the first JuiceStream should have a value of 0x is because the next object (in our case another JuiceStream) is on the same x coordinate, so it is directly above it. Instead it is 0.69x, which is the distance between the start of the first JuiceStream and the start of second JuiceStream.

For reference, the snapping values are also shown when selecting an object in stable on the top right corner of the editor: 
![image-29](https://github.com/user-attachments/assets/f467e36b-33d8-4605-a60c-84de6d308414)
In this case, in lazer only the "next" value is shown.

I hope this clarified it

**Reply by snalgae** [reply]
I kinda didn't realise the potential of the snapping bar until I found this issue here.

I want to clarify that lazer gives 1.65x snap for walks and 3.3x snap for dashes.

I would find having this snap useful to create specific distances of ratios of walks and dashes. If I want to create a distance that is .8x of a walk I could multiply 1.65x by it to get 1.32x snap. Or if I want a distance between a dash and a walk (1.5x) which would be roughly 2.48x snap. It is easier to measure out these ratios instead of measuring out pixels and better than placing it roughly there free-form, or trying to make ratios with the distance snap option.

There are some problems with snap bar that make it annoying (which I found):
- _The snap is based on the first object that is next on the timeline_
I find it annoying that I can't see the snap while placing or moving an object unless it's next in the timeline. I want to see the whole pattern to create an accurate placement for a snap instead of having to create the ratio first and then scroll up, or estimate it roughly on the spot and go back to see if it works.

- _1.65x is unnecessary to base calculations on_
I need to make more math than having a ratio in mind for a placement of an object and I think 1x is preferable (considering Spyreex assumed it was 1x which was intuitive to them, but it was wrong)

- _Snaps shown in the snapping bar of objects that go after a slider are based on the start of the slider instead of its end. (Spyreex's issue)_
This is a more visual example of the described issue

https://github.com/user-attachments/assets/0c0144b9-a6c1-4a9f-94af-e5feecf9b706


![Screenshot 2024-09-02 142241](https://github.com/user-attachments/assets/c343f52a-447c-4fcf-9712-6fa1aa840d44)

**Reply by peppy** [reply, marked as answer]
Thanks for the clear explanation! Will be fixed by https://github.com/ppy/osu/pull/29681.
